deep_water_level/src/deep_water_level/deeplabv3/train.py

```diff
@@ -0,0 +1,115 @@
+import time
+from pathlib import Path
+import os
+import torch
+import torch.nn as nn
+from annotation_utils.misc import my_device
+from deep_water_level.deeplabv3.utils import SegmentationDataset
+from torch.optim.adam import Adam
+from torch.utils.data import DataLoader
+from torchvision import models
+from torchvision.models.segmentation import DeepLabV3_ResNet50_Weights
+from torchvision.models.segmentation.deeplabv3 import DeepLabHead
+from torchvision.transforms import functional as F
+
+import mlflow
+
+
+def do_training(
+    # Dataset parameters
+    train_dataset_dir: Path,
+    test_dataset_dir: Path | None,
+    # Training parameters
+    n_epochs: int = 40,
+    batch_size: int = 8,
+    learning_rate: float = 1e-3,
+):
+    if test_dataset_dir is None or test_dataset_dir == train_dataset_dir:
+        # Split the train dataset in test and train datasets
+        dataset = SegmentationDataset(train_dataset_dir, train_dataset_dir / "annotations" / "filtered.json")
+
+        train_dataset, test_dataset = torch.utils.data.random_split(dataset, [0.8, 0.2])
+    else:
+        train_dataset = SegmentationDataset(train_dataset_dir, train_dataset_dir / "annotations" / "filtered.json")
+        test_dataset = SegmentationDataset(test_dataset_dir, test_dataset_dir / "annotations" / "filtered.json")
+
+    print(f"Training dataset size: {len(train_dataset)}, Test dataset size: {len(test_dataset)}")
+
+    device = my_device()
+
+    model = models.segmentation.deeplabv3_resnet50(weights=DeepLabV3_ResNet50_Weights.DEFAULT)
+    model.classifier = DeepLabHead(2048, num_classes=2)
+    model.to(device)
+
+    criterion = nn.CrossEntropyLoss()
```

Comment:
This is the loss because in the end it's classifying between different classes, right?

I wonder if it would be possible to adapt it to only work on two classes: water and "not water". In that case the criterion may need to be Binary Cross Entropy loss, right?